lib/matplotlib/tests/test_triangulation.py

```diff
@@ -1166,43 +1166,43 @@ def test_internal_cpp_api():
     # C++ Triangulation.
     with pytest.raises(
             TypeError,
-            match=r'function takes exactly 7 arguments \(0 given\)'):
+            match=r'__init__\(\): incompatible constructor arguments.'):
         mpl._tri.Triangulation()
 
     with pytest.raises(
             ValueError, match=r'x and y must be 1D arrays of the same length'):
-        mpl._tri.Triangulation([], [1], [[]], None, None, None, False)
+        mpl._tri.Triangulation([], [1], [[]], (), (), (), False)
```

Comment:
Because of lacking knowledge: will this cause problems for anyone possibly creating the Triangulation explicitly? Am I correct assuming that passing None is no longer feasible at this level?

Reply:
In principle yes, but we do not expose this class out publicly (there is a Python-side `Triangulation` class in `tri._triangulation` that we re-export in `tri/__init__.py`).

I think it is OK both that we change this API without warning and that we have tests that touch private API.

Reply:
I'm not entirely happy with this internal conversion of `None` to empty tuple for optional integer arrays but I cannot think of an alternative that isn't worse. In the long run we will use `std::optional` but that is C++17. I wouldn't want to use any C++17 until SciPy does, but that may not be far away (https://docs.scipy.org/doc/scipy/dev/toolchain.html#c-language-standards).

In the meantime we can use `std::optional` by shipping our own `optional.h` but I wouldn't want to do that as if there is a problem with some obscure compiler we will be on our own. Or we could write our own bespoke `np.array | None` pybind11 converter class but that is much more esoteric C++ than we currently have. So the least bad option is to stick with the `None` to tuple until we are OK with using C++17.

Note to self: need to add explicit internal API tests for `set_mask` as this also accepts an optional array.

Reply:
fwiw scipy seems to aim to bump to c++17 in the next (1.10) release, see https://github.com/scipy/scipy/pull/16589.